### ISSUE TITLE
Planes don't attack subs

### ISSUE THREAD

**Issue by beelee1** [triplea-game/triplea#7177]
### How can the problem be recreated?
Global 40 House Rules with Canada

Follow the prompts in the saved games. 13066 allows planes to shoot at subs but 20365 doesn't.

### Do you have any ideas for an expected fix?
See how 13066 did it.

### Attach a Save Game

[13066 planes do attack subs.zip](https://github.com/triplea-game/triplea/files/4928323/13066.planes.do.attack.subs.zip)

[20234 planes don't attack subs.zip](https://github.com/triplea-game/triplea/files/4928325/20234.planes.don.t.attack.subs.zip)

### If playing a prerelease, which version are you using?

Game Version:
2.1.20365

### If playing a prerelease, does this happen on the latest release?

- [] yes
- [] no

### Is there anything else we should know?

I turned on map option "Option Redesign" and turned off "Air Attack Sub Restricted". Scramble one ftr from Scotland. You may need to run more than once if sub is killed before ftr would shoot at it.

Trying to add the "Regression" label but haven't been able to figure it out.

**Comment by beelee1** [comment]
So I went through everything again and noticed this difference between 13066 and 20365:

![Screenshot from 2020-07-21 16-15-03](https://user-images.githubusercontent.com/24382682/88107862-7bf8d400-cb6d-11ea-899e-6a1023f69bd7.png)


![Screenshot from 2020-07-21 16-06-04](https://user-images.githubusercontent.com/24382682/88107479-c037a480-cb6c-11ea-9757-572809d21943.png)

It says"Air defend non subs" 20365 but not on 13066 which is the 2nd screenshot. You can also see where the ftr fired in the 13066 shot.

At any rate, this basically breaks how I have subs work and have been working on this mod for several years, so when someone gets time, I'd appreciate if you could make it work like 13066.

Thanks

**Comment by Cernelius** [comment]
@beelee1 Are you saying that air doesn't target submarines when "Air Attack Sub Restricted" is false since the start of the game or what?

**Comment by beelee1** [comment]
yea or what

With it is turned off planes get to shoot at subs w/o DD. Play the saves

**Comment by Cernelius** [comment]
I changed the WW2v3-1941.xml (the game is "World War II v3 1941") to have "Air Attack Sub Restricted" false and non editable.

With TripleA 2.1.20613, sending 1 Germans "fighter" to attack 1 Russians "submarine" in "4 Sea Zone", taking care the submarine never choosing to submerge, the "fighter" killed the "submarine" (so, no problem).

However, loading the provided "20234 planes don't attack subs" save-game with the same program, I confirm that, scrambling all available "fighters", they seem unable to target the "submarine", despite "Air Attack Sub Restricted" being false.

I don't know the reason of the different behaviour.

**Comment by beelee1** [comment]
@Cernelius Thank you for investigating it

**Comment by trevan** [comment]
The problem appears to be caused by the fact that when you change a "Map Option" (such as "Air Attack Sub Restricted"), it doesn't trigger the UnitAttachment's to update.  The submarine UnitAttachment sees "Air Attack Sub Restricted" as true when it is first loaded and so it sets up the submarine to be invisible to air units.

I don't know where in the code to trigger the rebuild of the UnitAttachment.  I'm not even sure if there is a place that does that.

If the option is false in the XML, then it will see that value when it first loads and work correctly.

**Comment by beelee1** [comment]
@trevan  Thanks for looking at it. I'ts just the fact that it worked in 13066 and doesn't now is the problem. Sorry idk how to do anything else to help.

**Comment by Cernelius** [comment]
> The problem appears to be caused by the fact that when you change a "Map Option" (such as "Air Attack Sub Restricted"), it doesn't trigger the UnitAttachment's to update. The submarine UnitAttachment sees "Air Attack Sub Restricted" as true when it is first loaded and so it sets up the submarine to be invisible to air units.
> 
> I don't know where in the code to trigger the rebuild of the UnitAttachment. I'm not even sure if there is a place that does that.
> 
> If the option is false in the XML, then it will see that value when it first loads and work correctly.

This is why I specifically asked if that setting was "false since the start of the game", not just "false" (and @beelee1 said that it was).

How do you see if and when a setting was changed after the start of the game? I believe that if you save the game, change settings and reload, there is no information about what were the previous settings.

**Comment by beelee1** [comment]
yea so the unitAttachment not updating would be what's different. I use a lot of map options, although this is the only problem I've noticed so far, but ... it will make play difficult if that's the case.

@Cernelius  so original game settings were all set before game started. Same as the saves. Sadly I can't find original save. Recreated to make it "easier" to see the problem.

Game is selected. Map options "Option Redesign" is selected and "Air Attack Sub Restricted" is turned off. This is the same for both 1.9.0.0.13066 and 2.1.20365. The behavior is different.

It is not being saved and changed in game

**Comment by trevan** [comment]
What should happen if "Option Redesign" is true and "Air Attack Sub Restricted" is true?

UnitAttachment has actually never been updated when a property changed (as far as I can tell).  The difference in 2.x is that `isSub` was redone and the properties are now being checked earlier in the game flow.  In 1.9, they were checked constantly during the course of the game.  Now they are only checked once.  And that check, unfortunately, is never redone even if you change the property.

**Comment by beelee1** [comment]
@trevan yea so the "isSub" thing is probably what's wrong.
Here's how subs and DDs work:

3) "SubsCanEvadeDestroyers_ChangerMustActivate" Destroyers no longer block Subs or the Sub's "First Strike" capability. They are now a A1 D1 M2 +1 w/NB C5 unit. They also have 1 preemptive "ASA" (Anti-Sub Attack) shot.

 Note If necessary activate "BB and CA 2 Shot" to prevent DDC5 spammage.

 When attacking a Sub, the Destroyer will fire a one time "ASA1" (hits at 2 out of 12) shot. If successful, the Sub is immediately destroyed and may not return fire. If the Sub survives, it may then submerge or take its normal "First Strike" shot.  

The Destroyer also has 1 preemptive "ASD" (Anti-Sub Defense) shot for combat and 1 for movement. It hits at 1 out of 12. So when a Sub attacks a Destroyer, it will undergo 1 ASD shot when the Sub enters the Destroyer's Sea Zone and 1 ASD shot at the beginning of Combat. Both of these attacks occur before the Subs First Strike.  

Destroyers will fire 1 ASD shot when Subs pass through the Destroyer's Sea Zone whether during Combat Move or Non Combat Move.  

Note When moving through or into Sea Zones containing a Destroyer, you should form your Subs into "Wolfpacks" and move them together where possible. The Destroyer only has 1 movement ASD shot and will only fire once. If you move into or through the Sea Zone separately, the Destroyer will fire each time a Sub enters it's Sea Zone. Destroyer ASA/D shots will stack. Meaning if there is only 1 Sub and multiple Destroyers, the sub will undergo as many ASA/D shots as there are Destroyers.  

Subs remain A2 D1 M2 +1 w/NB C6. Subs cannot be blocked by any Naval Vessel. Subs are now susceptible to counterattack from Fighters and Tac Bombers. When attacking naval units who have air units present, Subs will undergo counterfire from the air units regardless of the outcome of their "First Strike". When defending against attack by air units the sub may still submerge before the attack (except for the Tac Bombers ASA shot mentioned below).  

Subs in Blockade Zones may either retreat or undergo 1 Round of Combat from ASW Units during the Subs Combat phase.  

All other rules regarding Subs are unchanged.  

Note When moving a sub through a contested Sea Zone, you must move in one continuous movement. If you stop in the contested Sea Zone you will be unable to pass through.  

You will also be unable to undo your move once being "Depth Charged". Includes if you stop after the "Depth Charging" and still have movement left.  

In addition to it's normal unit stats, Tac Bombers now have 1 ASA1 and ASD1 each. Their ASA shot will hit at 2 out of 12. Their ASD shot will hit at 1 out of 12. They will stack. Meaning if there is only 1 Sub and multiple Tacs, the sub will undergo as many ASA/ASD shots as there are Tacs. They only fire during the Combat Phase.  

As mentioned above, it now fires normally at subs without a Destroyer present.  

Note If "PlanesTargetNaval" and or "PlanesTargetNavalCV" has been activated then Tactical Bombers ASA and ASD shots will no longer Stack. Meaning only 1 shot per sub no matter how many Tacs are present.   

May also be activated via Map Options.

So idk i can't find the original game where it happened, but it's when subs attack and defensive planes are present, they can't hit the planes but the plane can hit the subs. If planes attack subs, the subs can submerge unless they have AA attack on it, then they can submerge, if they survive, before regular combat.

So basically, if they attack and the "Air Attack Sub" option is off, they should get shot at.

**Comment by trevan** [comment]
I created a PR to fix this.  If you know how to build from source, it would be nice if you could test it.  The save games will not be fixed, though they will load.  You'll have to restart the game.  It looks like the game is correctly rolling for the airplanes now.

**Comment by trevan** [comment]
The PR won't work because of saved games and triggers.

**Comment by Cernelius** [comment]
@trevan Are you sure that this is a regression, in the moment that (amongst other things) the fact that "Air Attack Sub Restricted" would now have this behaviour was wanted by at least @ron-murhammer? Maybe this is more of a matter of missing or unclear deprecation?

**Comment by trevan** [comment]
> @trevan Are you sure that this is a regression, in the moment that (amongst other things) the fact that "Air Attack Sub Restricted" would now have this behaviour was wanted by at least @ron-murhammer? Maybe this is more of a matter of missing or unclear deprecation?

Yes, it is a regression.  If the map has the property set to false, then it will act as @beelee1 expects.  The bug manifests itself only when you make the property editable and set it to false in the UI before you start the game.  So the property works as expected.  It is just the change to the property (from true -> false) in the UI after the map has been loaded that is not being recognized.  And that is because in 1.9, the property was checked every single time it needed to be checked.  In 2.0, it is only checked once when the map is loaded.

**Comment by Cernelius** [comment]
> > @trevan Are you sure that this is a regression, in the moment that (amongst other things) the fact that "Air Attack Sub Restricted" would now have this behaviour was wanted by at least @ron-murhammer? Maybe this is more of a matter of missing or unclear deprecation?
> 
> Yes, it is a regression. If the map has the property set to false, then it will act as @beelee1 expects. The bug manifests itself only when you make the property editable and set it to false in the UI before you start the game. So the property works as expected. It is just the change to the property (from true -> false) in the UI after the map has been loaded that is not being recognized. And that is because in 1.9, the property was checked every single time it needed to be checked. In 2.0, it is only checked once when the map is loaded.

I mean, I would totally be with you not liking this behaviour and considering it wrong, and I said so pretty much since ever (and I've always been alone at that, so now I guess we are two). I'm just letting you know that what you describe is exactly what both @ron-murhammer and whoever merged its PR, when this change was made, wanted, at least from what I can understand from past discussions like:
https://forums.triplea-game.org/topic/328/unit-option-can-submerge-hide-for-land-units-partisan-guerrilla-spy-diplomat-munition/46?page=3
>I understand what you are getting at but its much simpler to minimize the usage of some of these properties in the engine and have them influence setting the new unit options then setting all new unit options to true and then also check the property. You really wouldn't ever be changing these type of sub properties part way through a game anyways. There are just too many confusing sub properties that very very few people understand so doing whatever I can to minimize them helps.

**Comment by trevan** [comment]
I mentioned this in my closed PR but this issue affects at least 4 properties:
* "Air Attack Sub Restricted"
* "Submersible Subs"
* "Ignore Sub In Movement"
* "Defending Suicide and Munition Units Do Not Fire"

If you have a map that has those properties set as editable, then when you edit the property in the UI, the change will not take affect.

**Comment by Cernelius** [comment]
> I mentioned this in my closed PR but this issue affects at least 4 properties:
> 
> * "Air Attack Sub Restricted"
> * "Submersible Subs"
> * "Ignore Sub In Movement"
> * "Defending Suicide and Munition Units Do Not Fire"
> 
> If you have a map that has those properties set as editable, then when you edit the property in the UI, the change will not take affect.

That's because @ron-murhammer is doing it that way, generally. That was my point, as well (I made over 1 year ago):
https://forums.triplea-game.org/topic/328/unit-option-can-submerge-hide-for-land-units-partisan-guerrilla-spy-diplomat-munition/44?page=3
>Generally speaking, I think that fluid/editable (and currently not even validated!) xml elements, like the properties are, should never be taken as pseudostatic references to define what is ultimately burned into the savegame

---

My question was essentially whether or not something can be considered a regression if it works exactly as intended (as I believe it does, as much as I disagree with the intention, to start with). Anyway, let's drop this matter: It's just a label.

**Comment by beelee1** [comment]
@trevan Thank you. I haven't built from source before. I'll give it a try

**Comment by beelee1** [comment]
@trevan I remember trying to "build from source" a while back but didn't pursue it much. I thought I had saved some info about it but can't find it. Can you post a link to get me started ?

I found this do i need java 11 ?

https://forums.triplea-game.org/topic/1437/how-to-build-from-source-for-dummies/2?_=1595628957020

"But if you really insist on building triplea from source you can of course do that as well by checking out repository via git and running ./gradlew :game-headed:run, which should build it from source and execute it right away when you have java 11/8 installed on your system."

 What java software do i need to do the build ?

Thanks

**Comment by trevan** [comment]
@beelee1 , I have openjdk 11.

How I build it is using IntelliJ.  You can download it from https://www.jetbrains.com/idea/.  Then you'll want to checkout the source code and follow the steps https://github.com/triplea-game/triplea/wiki/IntelliJ-Setup on setting it up.  After that, in the top rightish corner, there is a way to run the "HeadedGameRunner".

Eclipse is also an option.  The steps for that is https://github.com/triplea-game/triplea/wiki/Eclipse-Setup.

**Comment by beelee1** [comment]
@trevan thanks again. Appreciate your time

**Comment by beelee1** [comment]
@trevan 
Having some trouble installing idea. Here's what i have:

$ ./idea.sh
OpenJDK 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
Jul 24, 2020 7:37:11 PM java.util.prefs.FileSystemPreferences$6 run
WARNING: Prefs file removed in background /home/greg/.java/.userPrefs/prefs.xml
2020-07-24 19:39:16,452 [ 124792]  ERROR - tellij.diagnostic.LoadingState - Should be called at least in the state CONFIGURATION_STORE_INITIALIZED, the current state is: LAF_INITIALIZED
Current violators count: 1

 
java.lang.Throwable
	at com.intellij.diagnostic.LoadingState.logStateError(LoadingState.java:56)
.
.
"bunch of lines similar"
.
.
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
2020-07-24 19:39:16,457 [ 124797]  ERROR - tellij.diagnostic.LoadingState - IntelliJ IDEA 2020.1.4  Build #IC-201.8743.12 
2020-07-24 19:39:16,457 [ 124797]  ERROR - tellij.diagnostic.LoadingState - JDK: 11.0.7; VM: OpenJDK 64-Bit Server VM; Vendor: JetBrains s.r.o. 
2020-07-24 19:39:16,457 [ 124797]  ERROR - tellij.diagnostic.LoadingState - OS: Linux 
2020-07-24 19:39:21,403 [ 129743]   WARN - Container.ComponentManagerImpl - Do not use constructor injection (requestorClass=com.android.tools.idea.sdk.AndroidSdks) 


"stopped until ^C"


C2020-07-24 19:47:53,814 [ 642154]   WARN - pl.local.NativeFileWatcherImpl - Watcher terminated with exit code 130

**Comment by DanVanAtta** [comment]
@beelee1 I get a number of warnings when launching, most of those are probably expected.

When you installed Idea, did it give you an option to create a desktop icon or a launcher in /bin? Ideally you can just type `idea` and the installed wrapper scripts will kick in. Perhaps try re-installing with those options.

Failing using idea, you'll need to clone the TripleA project and perhaps check out the experimental branch that is being worked on. If you can get that far, then running `./gradlew :game-headed:run` in the TripleA project folder will also build from source (gradlew will download all needed dependencies, it should "just work")
